Invalid pattern sent in the configuration shuts down the webapp

Invalid pattern (should be `.*rok` and is `*rok`) produces an unchecked exception and the whole webapp fails and stops when tries to work with that pattern after.

Receiving message (this is quite ok):
```log
22-Feb-2017 11:49:45.934 WARNING [configurationListener] org.opensolaris.opengrok.configuration.RuntimeEnvironment.handleMessage Message dropped: {0} - message error
 java.util.regex.PatternSyntaxException: Dangling meta character '*' near index 1
(*rok)
 ^
	at java.util.regex.Pattern.error(Pattern.java:1955)
	at java.util.regex.Pattern.sequence(Pattern.java:2123)
	at java.util.regex.Pattern.expr(Pattern.java:1996)
	at java.util.regex.Pattern.group0(Pattern.java:2905)
	at java.util.regex.Pattern.sequence(Pattern.java:2051)
	at java.util.regex.Pattern.expr(Pattern.java:1996)
	at java.util.regex.Pattern.compile(Pattern.java:1696)
	at java.util.regex.Pattern.<init>(Pattern.java:1351)
	at java.util.regex.Pattern.compile(Pattern.java:1028)
	at java.util.regex.Pattern.matches(Pattern.java:1133)
	at org.opensolaris.opengrok.configuration.Group.match(Group.java:154)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment$5.test(RuntimeEnvironment.java:1238)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment$5.test(RuntimeEnvironment.java:1235)
	at java.util.Collection.removeIf(Collection.java:414)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.populateGroups(RuntimeEnvironment.java:1235)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.setConfiguration(RuntimeEnvironment.java:1270)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.applyConfig(RuntimeEnvironment.java:1565)
	at org.opensolaris.opengrok.configuration.messages.ConfigMessage.applyMessage(ConfigMessage.java:39)
	at org.opensolaris.opengrok.configuration.messages.Message.apply(Message.java:85)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.handleMessage(RuntimeEnvironment.java:1668)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment$10.run(RuntimeEnvironment.java:1628)
	at java.lang.Thread.run(Thread.java:745)
```

**However the message was applied half way so the group is already there in the environment with the invalid pattern!!**

Tomcat log (webapp failed):
```log
22-Feb-2017 11:55:32.357 SEVERE [http-nio-8080-exec-19] org.apache.catalina.core.StandardHostValve.custom Exception Processing ErrorPage[errorCode=500, location=/error]
 org.apache.jasper.JasperException: An exception occurred processing JSP page /menu.jspf at line 52

49: /* ---------------------- menu.jspf start --------------------- */
50: {
51:     PageConfig cfg = PageConfig.get(request);
52:     ProjectHelper ph = ProjectHelper.getInstance(cfg);
53:     JSONArray messages;
54:     Set<Project> projects = ph.getAllProjects();
55:     if (projects == null) {


Stacktrace:
	at org.apache.jasper.servlet.JspServletWrapper.handleJspException(JspServletWrapper.java:574)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:476)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:396)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:340)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:721)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:468)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:391)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:318)
	at org.apache.catalina.core.StandardHostValve.custom(StandardHostValve.java:439)
	at org.apache.catalina.core.StandardHostValve.status(StandardHostValve.java:305)
	at org.apache.catalina.core.StandardHostValve.throwable(StandardHostValve.java:399)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:180)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1526)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1482)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.regex.PatternSyntaxException: Dangling meta character '*' near index 1
(*rok)
 ^
	at java.util.regex.Pattern.error(Pattern.java:1955)
	at java.util.regex.Pattern.sequence(Pattern.java:2123)
	at java.util.regex.Pattern.expr(Pattern.java:1996)
	at java.util.regex.Pattern.group0(Pattern.java:2905)
	at java.util.regex.Pattern.sequence(Pattern.java:2051)
	at java.util.regex.Pattern.expr(Pattern.java:1996)
	at java.util.regex.Pattern.compile(Pattern.java:1696)
	at java.util.regex.Pattern.<init>(Pattern.java:1351)
	at java.util.regex.Pattern.compile(Pattern.java:1028)
	at java.util.regex.Pattern.matches(Pattern.java:1133)
	at org.opensolaris.opengrok.configuration.Group.match(Group.java:154)
	at org.opensolaris.opengrok.web.ProjectHelper$1.test(ProjectHelper.java:132)
	at org.opensolaris.opengrok.web.ProjectHelper$1.test(ProjectHelper.java:129)
	at java.util.Collection.removeIf(Collection.java:414)
	at org.opensolaris.opengrok.web.ProjectHelper.populateGroups(ProjectHelper.java:129)
	at org.opensolaris.opengrok.web.ProjectHelper.<init>(ProjectHelper.java:86)
	at org.opensolaris.opengrok.web.ProjectHelper.getInstance(ProjectHelper.java:100)
	at org.apache.jsp.error_jsp._jspService(error_jsp.java:288)
	at org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:438)
	... 25 more

22-Feb-2017 11:56:40.155 SEVERE [http-nio-8080-exec-22] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class org.opensolaris.opengrok.web.WebappListener
 java.util.regex.PatternSyntaxException: Dangling meta character '*' near index 1
(*rok)
 ^
	at java.util.regex.Pattern.error(Pattern.java:1955)
	at java.util.regex.Pattern.sequence(Pattern.java:2123)
	at java.util.regex.Pattern.expr(Pattern.java:1996)
	at java.util.regex.Pattern.group0(Pattern.java:2905)
	at java.util.regex.Pattern.sequence(Pattern.java:2051)
	at java.util.regex.Pattern.expr(Pattern.java:1996)
	at java.util.regex.Pattern.compile(Pattern.java:1696)
	at java.util.regex.Pattern.<init>(Pattern.java:1351)
	at java.util.regex.Pattern.compile(Pattern.java:1028)
	at java.util.regex.Pattern.matches(Pattern.java:1133)
	at org.opensolaris.opengrok.configuration.Group.match(Group.java:154)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment$5.test(RuntimeEnvironment.java:1238)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment$5.test(RuntimeEnvironment.java:1235)
	at java.util.Collection.removeIf(Collection.java:414)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.populateGroups(RuntimeEnvironment.java:1235)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.setConfiguration(RuntimeEnvironment.java:1270)
	at org.opensolaris.opengrok.configuration.RuntimeEnvironment.readConfiguration(RuntimeEnvironment.java:1139)
	at org.opensolaris.opengrok.web.WebappListener.contextInitialized(WebappListener.java:74)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4729)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5167)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:725)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:701)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:717)
	at org.apache.catalina.startup.HostConfig.deployDescriptor(HostConfig.java:586)
	at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:460)
	at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1556)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:300)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
	at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at org.apache.catalina.manager.ManagerServlet.check(ManagerServlet.java:1460)
	at org.apache.catalina.manager.ManagerServlet.deploy(ManagerServlet.java:906)
	at org.apache.catalina.manager.ManagerServlet.doGet(ManagerServlet.java:344)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.filters.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:108)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:614)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1526)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1482)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/src/org/opensolaris/opengrok/configuration/Group.java
+++ b/src/org/opensolaris/opengrok/configuration/Group.java
@@ -26,9 +26,12 @@ import java.util.Locale;
 import java.util.Set;
 import java.util.TreeSet;
 import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
 
 /**
- * Placeholder for the information about subgroups of projects and repositories
+ * Placeholder for the information about subgroups of projects and repositories.
+ *
+ * Supports natural ordering based on case insensitive group names.
  *
  * @author Krystof Tulinger
  * @version $Revision$
@@ -36,7 +39,26 @@ import java.util.regex.Pattern;
 public class Group implements Comparable<Group>, Nameable {
 
     private String name;
+    /**
+     * Group regexp pattern.
+     *
+     * No project matches the empty pattern of "" however this group can still
+     * be used as a superior group for other groups (without duplicating the
+     * projects).
+     */
     private String pattern = "";
+    /**
+     * Compiled group pattern.
+     *
+     * We set up the empty compiled pattern by default to "()" to reduce code
+     * complexity when performing a match for a group without a pattern.
+     *
+     * This pattern is updated whenever the string pattern {@link #pattern} is
+     * updated.
+     *
+     * @see #setPattern(String)
+     */
+    private Pattern compiledPattern = Pattern.compile("()");
     private Group parent;
     private int flag;
 
@@ -99,8 +121,8 @@ public class Group implements Comparable<Group>, Nameable {
         subgroups.add(g);
         descendants.add(g);
     }
-    
-    public Set<Group> getParents () {
+
+    public Set<Group> getParents() {
         if (parents == null) {
             parents = new TreeSet<>();
             Group tmp = parent;
@@ -134,7 +156,14 @@ public class Group implements Comparable<Group>, Nameable {
         return pattern;
     }
 
-    public void setPattern(String pattern) {
+    /**
+     * Set the group pattern.
+     *
+     * @param pattern the regexp pattern for this group
+     * @throws PatternSyntaxException when the pattern is invalid
+     */
+    public void setPattern(String pattern) throws PatternSyntaxException {
+        this.compiledPattern = Pattern.compile("(" + pattern + ")");
         this.pattern = pattern;
     }
 
@@ -153,7 +182,7 @@ public class Group implements Comparable<Group>, Nameable {
      * @return true if project's description matches the group pattern
      */
     public boolean match(Project p) {
-        return Pattern.matches("(" + this.pattern + ")", p.getName());
+        return compiledPattern.matcher(p.getName()).matches();
     }
 
     @Override

--- a/test/org/opensolaris/opengrok/configuration/GroupTest.java
+++ b/test/org/opensolaris/opengrok/configuration/GroupTest.java
@@ -29,6 +29,7 @@ import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.util.LinkedList;
 import java.util.TreeSet;
+import java.util.regex.PatternSyntaxException;
 import junit.framework.AssertionFailedError;
 import org.junit.Test;
 
@@ -81,6 +82,33 @@ public class GroupTest {
             // Can only chain one of the exceptions. Take the first one.
             afe.initCause(exceptions.getFirst());
             throw afe;
+        }
+    }
+
+    @Test
+    public void invalidPatternTest() {
+        testPattern("*dangling asterisk", false);
+        testPattern(".*(", false);
+        testPattern("+", false);
+        testPattern("[a-z?.*", false);
+        testPattern("()", true);
+        testPattern("[a-z?(.*)]", true);
+        testPattern("[a-z?.*]", true);
+        testPattern("valid pattern", true);
+        testPattern(".*(.*.*)?\\*.*", true);
+    }
+
+    private void testPattern(String pattern, boolean valid) {
+        try {
+            Group g = new Group();
+            g.setPattern(pattern);
+            if (!valid) {
+                fail("Pattern \"" + pattern + "\" is invalid regex pattern, exception expected.");
+            }
+        } catch (PatternSyntaxException ex) {
+            if (valid) {
+                fail("Pattern \"" + pattern + "\" is valid regex pattern, exception thrown.");
+            }
         }
     }
 

